packages/nightingale-track-canvas/src/utils/utils.ts

```diff
@@ -0,0 +1,208 @@
+ * Not suited for storing float intervals, but query interval can be float. */
+export class RangeCollection<T> {
+    protected readonly items: T[];
+    protected readonly starts: number[];
+    protected readonly stops: number[];
+
+    /** Keys to `this.bins`, sorted in ascending order */
+    protected readonly binSpans: number[];
+    /** `this.bins[span]` contains indices of all items whose length is `<= span` but `> span/Q`, in ascending order */
+    protected readonly bins: Record<number, number[]>;
+    /** Ratio of spans of neighboring bins */
+    protected readonly Q = 2;
+    /** Reusable arrays (one for each bin), to avoid repeated array allocation */
+    private readonly _tmpArrays: Record<number, number[]> = {};
+
+    /** Create a new collection of ranges. `start` must return range start (inclusive), `stop` must return range end (exclusive) */
+    constructor(items: T[], accessors: { start: (item: T) => number, stop: (item: T) => number }) {
+        const { start, stop } = accessors;
+        this.items = items;
+        this.starts = items.map(start);
+        this.stops = items.map(stop);
+
+        this.bins = {};
+        for (let i = 0; i < items.length; i++) {
+            const length = this.stops[i] - this.starts[i];
+            let binSpans = 1;
+            while (binSpans < length) binSpans *= this.Q;
+            (this.bins[binSpans] ??= []).push(i);
+        }
+        this.binSpans = sortNumeric(Object.keys(this.bins).map(Number));
+        for (const binSpan of this.binSpans) {
+            this.bins[binSpan].sort(this.compareFn);
+        }
+    }
+
+    /** Return number of items. */
+    size(): number {
+        return this.items.length;
+    }
+
+    /** Get all ranges that overlap with interval [start, stop).
+     * Does not preserve original order of the ranges!
+     * Instead sorts the ranges by their start (ranges with the same start are sorted by decreasing length). */
+    overlappingItems(start: number, stop: number): T[] {
+        return this.overlappingItemIndices(start, stop).map(i => this.items[i]);
+    }
+
+    /** Get indices of all ranges that overlap with interval [start, stop).
+     * Does not preserve original order of the ranges!
+     * Instead sorts the ranges by their start (ranges with the same start are sorted by decreasing length). */
+    overlappingItemIndices(start: number, stop: number): number[] {
+        const partialOuts = this.binSpans.map(binSpan => this.overlappingItemIndicesInBin(binSpan, start, stop, this._tmpArrays[binSpan] ??= []));
+        return mergeSortedArrays(partialOuts, this.compareFn);
+    }
+
+    private overlappingItemIndicesInBin(binSpan: number, start: number, stop: number, out: number[]): number[] {
+        out.length = 0;
+        const bin = this.bins[binSpan];
+        const from = firstGteqIndex(bin, Math.floor(start) - binSpan + 1, i => this.starts[i]);
```

Comment:
Not sure I understand why `Math.floor(start) - binSpan + 1` is done?

Reply:
Items in that bin have length <=binSpan, so item_stop <= item_start + binSpan. 
Overlapping items must fulfil: 
    item_stop > start
Which can be transformed to:
    item_start + binSpan > start 
    item_start > start - binSpan
And the items have integer start and end, so:
    item_start >= Math.floor(start) - binSpan + 1
    
But now that you pointed out, the last step is not really necessary and it just complicate things. I will simplify it. Thanks